forecast-bbs-core.R

```diff
@@ -28,35 +28,75 @@ install_dataset <- function(dataset){
 #'
 #' Removes waterbirds, shorebirds, owls, kingfishers, knightjars,
 #' dippers. These species are poorly sampled due to their aquatic or
-#' noctural nature. Also removes taxa that were either partially unidentified 
+#' noctural nature. Also removes taxa that were either partially unidentified
 #' (e.g. "sp.") or were considered hybrids (e.g. "A x B").
 #'
 #' @param df dataframe containing an species_id column
 #'
 #' @return dataframe, filtered version of initial dataframe
 filter_species <- function(df){
-  
-  con <- dbConnect(RPostgres::Postgres(), dbname = 'postgres')
-  species_table = dbFetch(dbSendQuery(con, "SELECT * FROM bbs.species;"))
-  
-  
-  unidentified = function(names) {
+  species_table = get_species_data()
+
+  is_unidentified = function(names) {
     grepl("/|unid\\.|sp\\.| or |hybrid| X | x ", names)
   }
-  
-  identified_taxa = species_table %>%
-    filter(!unidentified(english_common_name)) %>%
-    filter(!unidentified(spanish_common_name)) %>%
+
+  valid_taxa = species_table %>%
+    filter(!is_unidentified(english_common_name)) %>%
+    filter(!is_unidentified(spanish_common_name)) %>%
+    filter(aou > 2880) %>%
+    filter(aou < 3650 | aou > 3810) %>%
+    filter(aou < 3900 | aou > 3910) %>%
+    filter(aou < 4160 | aou > 4210) %>%
+    filter(aou != 7010)
+
+  filter(df, species_id %in% valid_taxa$aou)
+}
+
+combine_subspecies = function(df){
+
+  species_table = get_species_data()
+
+  # Subspecies have two spaces separated by non-spaces
+  subspecies_names = species_table %>%
+    filter(species_table$aou %in% unique(df$species_id)) %>%
+    magrittr::extract2("spanish_common_name") %>%
+    grep(" [^ ]+ ", ., value = TRUE)
+
+  subspecies_ids = species_table %>%
+    filter(spanish_common_name %in% subspecies_names) %>%
+    magrittr::extract2("aou")
+
+  # Drop the third word of the subspecies name to get the species name,
+  # then find the AOU code
+  new_subspecies_ids = species_table %>%
+    slice(match(gsub(" [^ ]+$", "", subspecies_names),
+                species_table$spanish_common_name)) %>%
     magrittr::extract2("aou")
-  
-  
+
+  # replace the full subspecies names with species-level names
+  for (i in seq_along(subspecies_ids)) {
+    df$species_id[df$species_id == subspecies_ids[i]] = new_subspecies_ids[i]
+  }
+
   df %>%
-    filter(species_id > 2880) %>%
-    filter(species_id < 3650 | species_id > 3810) %>%
-    filter(species_id < 3900 | species_id > 3910) %>%
-    filter(species_id < 4160 | species_id > 4210) %>%
-    filter(species_id != 7010) %>%
-    filter(species_id %in% identified_taxa)
+    filter(species_id %in% new_subspecies_ids) %>%
+    group_by(site_id, year, species_id) %>%
+    summarize(abundance = sum(abundance)) %>%
+    left_join(dplyr::select(df, -abundance),
+              by = c("site_id", "year", "species_id"))
+}
```

Comment:
ok that makes more sense. But now that I run it, the total rows in df isn't being reduced like it should be from aggregating species. It looks like the left join is adding things back in somehow. I think we can just get rid of the join and put  lat, long in the `group_by()` part to retain them. lat longs are unique to each site_id, so it's not adding "extra groups". 

df %>%
    group_by(site_id, year, species_id, lat, long) %>%
    summarize(abundance = sum(abundance)) %>%
    ungroup()